lib/Target/ExportSystemC/Patterns/SystemCEmissionPatterns.cpp

```diff
@@ -68,6 +68,109 @@ struct BuiltinModuleEmitter : OpEmissionPattern<ModuleOp> {
     p.emitRegion(op.getRegion(), scope);
   }
 };
+
+/// Emit a systemc.signal.write operation by using the explicit 'write' member
+/// function of the signal and port classes.
+struct SignalWriteEmitter : OpEmissionPattern<SignalWriteOp> {
+  using OpEmissionPattern::OpEmissionPattern;
+
+  void emitStatement(SignalWriteOp op, EmissionPrinter &p) override {
+    p.getInlinable(op.getDest()).emit();
+    p << ".write(";
+    p.getInlinable(op.getSrc()).emit();
+    p << ");\n";
+  }
+};
+
+/// Emit a systemc.signal.read operation by using the explicit 'read' member
+/// function of the signal and port classes.
+struct SignalReadEmitter : OpEmissionPattern<SignalReadOp> {
+  using OpEmissionPattern::OpEmissionPattern;
+
+  MatchResult matchInlinable(Value value) override {
+    if (value.getDefiningOp<SignalReadOp>())
+      return Precedence::FUNCTION_CALL;
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    p.getInlinable(value.getDefiningOp<SignalReadOp>().getInput()).emit();
+    p << ".read()";
+  }
+};
+
+/// Emit a systemc.ctor operation by using the SC_CTOR macro.
+struct CtorEmitter : OpEmissionPattern<CtorOp> {
+  using OpEmissionPattern::OpEmissionPattern;
+
+  void emitStatement(CtorOp op, EmissionPrinter &p) override {
+    p << "\nSC_CTOR(" << op->getParentOfType<SCModuleOp>().getModuleName()
+      << ") ";
+    p.emitRegion(op.getBody());
+  }
+};
+
+/// Emit a systemc.func operation.
+struct SCFuncEmitter : OpEmissionPattern<SCFuncOp> {
+  using OpEmissionPattern::OpEmissionPattern;
+
+  MatchResult matchInlinable(Value value) override {
+    if (value.getDefiningOp<SCFuncOp>())
+      return Precedence::VAR;
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    p << value.getDefiningOp<SCFuncOp>().getName();
+  }
+
+  void emitStatement(SCFuncOp op, EmissionPrinter &p) override {
+    p << "\nvoid " << op.getName() << "() ";
```

Comment:
There is some inconsistent usage of newlines before `emitStatement` and after `emitStatement`. Is there a way to make this more uniform/predictable? Can this be handled instead by the infrastructure automatically?

Reply:
Patterns have to emit a newline at the end of the statement(s) they emit. There is only one exception: when they call `emitRegion` without passing a custom scope. This is because it saves an additional `p << "/n";` at the end of the module, ctor, and func emitters and you probably always want a newline after a scope end. The additional newlines at the beginning of the patterns are just to ensure an empty line to improve readability (humans typically also add an empty line before every class or function signature). If we want the infrastructure to emit the newline after a statement, we would have to tell it somehow (e.g., by letting `emitStatement` return a boolean) whether the op actually emitted something (because ops that are always inlined don't emit anything in `emitStatement`) or otherwise we would end up with a lot of semi-random empty lines.

I added some comments that state the reason for the newlines at the start of the patterns.